### PR TITLE
feat: Proposed fix dialog

### PR DESCRIPTION
**Pull request by mclarke47**
## Hey, I just made a Pull Request!

Fixes https://github.com/backstage/backstage/issues/17136

Screenshot 2023-04-29 at 11 21 17 AM
Screenshot 2023-04-29 at 11 20 42 AM
Screenshot 2023-04-29 at 10 58 28 AM

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by mclarke47** [comment]
Waiting on https://github.com/backstage/backstage/pull/17563

**Comment by benjdlambert** [comment]
@jamieklassen do you wanna have a flick through this? :pray:

**Comment by mclarke47** [comment]
@jamieklassen and I will review in a meeting on thursday

**Comment by benjdlambert** [comment]
@mclarke47 nice!

File: plugins/kubernetes/src/api/KubernetesProxyClient.ts [hunk 1]
```diff
@@ -41,20 +42,63 @@ export class KubernetesProxyClient {
     return await response.text();
   }
 
+  private async handleResponse(response: Response): Promise<any> {
```
Comment (jamieklassen) [inline plugins/kubernetes/src/api/KubernetesProxyClient.ts:45]:
```suggestion
  private async handleJson(response: Response): Promise<any> {
```

File: plugins/kubernetes/src/api/KubernetesProxyClient.ts [hunk 2]
```diff
@@ -41,20 +42,63 @@ export class KubernetesProxyClient {
...
     clusterName: string;
     containerName: string;
+    previous?: boolean;
   }): Promise<{ text: string }> {
     const params = new URLSearchParams({
       container: containerName,
     });
+    if (previous) {
+      params.append('previous', '');
+    }
```
Comment (jamieklassen) [inline plugins/kubernetes/src/api/KubernetesProxyClient.ts:101]: Can we have a unit test for this?

File: plugins/kubernetes/src/components/Pods/Events/Events.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+          {events
+            .filter(event => {
+              if (warningEventsOnly) {
+                return event.type === 'Warning';
+              }
+              return true;
+            })
+            .map(event => {
+              const timeAgo = DateTime.fromISO(
+                event.metadata.creationTimestamp as any,
```
Comment (mclarke47) [inline plugins/kubernetes/src/components/Pods/Events/Events.tsx:83]: Does this need `as any`

File: plugins/kubernetes/src/components/Pods/Events/useEvents.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  namespace: string;
+  clusterName: string;
+}
+
+/**
+ * Retrieves the events for the given object
+ *
+ * @public
+ */
+export const useEvents = ({
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/Pods/Events/useEvents.ts:36]: can we unit test the hook?

File: plugins/kubernetes/src/components/Pods/FixDialog/FixDialog.test.tsx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            namespace: 'some-namespace',
+            kind: 'Pod',
+            apiGroup: 'v1',
+          },
+          proposedFix: {
+            type: 'logs',
+            container: 'some-container',
+            errorType: 'some error type',
+            rootCauseExplanation: 'some root cause',
+            possibleFixes: ['fix1', 'fix2'],
```
Comment (mclarke47) [inline plugins/kubernetes/src/components/Pods/FixDialog/FixDialog.test.tsx:136]: Rename possibleFixes to actions

File: plugins/kubernetes/src/components/Pods/PodLogs/PodLogsDialog.tsx [hunk 6]
```diff
@@ -43,11 +43,23 @@ const useStyles = makeStyles((theme: Theme) =>
...
 );
 
-interface PodLogsDialogProps {
+/**
+ * Props for PodLogsDialog
+ *
+ * @public
+ */
+export interface PodLogsDialogProps {
   podScope: ContainerScope;
```
Comment (mclarke47) [inline plugins/kubernetes/src/components/Pods/PodLogs/PodLogsDialog.tsx:52]: missed rename

File: plugins/kubernetes/src/error-detection/pods.ts [hunk 7]
```diff
@@ -70,6 +70,132 @@ const podToContainerSpecsAndStatuses = (pod: Pod): ContainerSpecAndStatus[] => {
...
+
+  return {
+    errorType: 'ReadinessProbeFailed',
+    rootCauseExplanation: `The container ${firstUnreadyContainerStatus?.name} failed to start properly, but is not crashing`,
+    possibleFixes: [
+      'Ensure that the container starts correctly locally',
+      "Check the container's logs looking for error during startup",
+    ],
+    type: 'events',
+    docsLink: 'TODO', // TODO not implement yet
```
Comment (mclarke47) [inline plugins/kubernetes/src/error-detection/pods.ts:88]: Event proposed fixes don't need docslink

File: plugins/kubernetes/src/components/Pods/FixDialog/FixDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+            {pf && pf.type === 'events' && (
+              <Button
+                disabled={hasPodEvents}
+                variant="outlined"
+                startIcon={<SubjectIcon />}
+                onClick={() => setHasPodEvents(true)}
+              >
+                Pod Events
+              </Button>
+            )}
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/Pods/FixDialog/FixDialog.tsx:187]: are these buttons redundant if you've already got the important info in the dialog content?

File: plugins/kubernetes/src/components/Pods/FixDialog/FixDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        {hasPodEvents && pf && pf.type === 'events' && (
+          <Grid item>
+            <Events
+              warningEventsOnly
+              involvedObjectName={pod.metadata?.name ?? ''}
+              namespace={pod.metadata?.namespace ?? ''}
+              clusterName={clusterName}
+            />
+          </Grid>
+        )}
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/Pods/FixDialog/FixDialog.tsx:135]: can this extra content have a title?

Comment (mclarke47) [reply plugins/kubernetes/src/components/Pods/FixDialog/FixDialog.tsx:135]: fixed

File: plugins/kubernetes/src/components/Pods/FixDialog/FixDialog.tsx (hunk 8 above)
Comment (mclarke47) [reply plugins/kubernetes/src/components/Pods/FixDialog/FixDialog.tsx:187]: fixed, removed them

File: plugins/kubernetes/src/components/Pods/Events/useEvents.ts (hunk 4 above)
Comment (mclarke47) [reply plugins/kubernetes/src/components/Pods/Events/useEvents.ts:36]: fixed

File: plugins/kubernetes/src/api/KubernetesProxyClient.ts (hunk 2 above)
Comment (mclarke47) [reply plugins/kubernetes/src/api/KubernetesProxyClient.ts:101]: fixed

File: plugins/kubernetes/src/api/KubernetesProxyClient.ts (hunk 1 above)
Comment (mclarke47) [reply plugins/kubernetes/src/api/KubernetesProxyClient.ts:45]: fixed

**Comment by mclarke47** [comment]
@jamieklassen ready for review again

**Review by jamieklassen — APPROVED** [review]

**Pull request merged**